8301261: Add linker option for specifying uncaught exception handler

JornVernee opened the pull request:
This patch adds the ability to specify an uncaught exception handler for an upcall stub, and changes the uncaught exception handling code to delegate to the thread's installed uncaught exception handler when no handler is specified as well, which will also allow setting a system-wide uncaught exception handler.

For the tests: I've refactored TestUpcallException a bit. Previously the test code was found in a separate class (ThrowingUpcall). I've removed this class and moved the code into nested classes of TestUpcallException instead. Then I've added new test cases for testing the uncaught exception handler.

I also noticed that we were setting the wrong system property to control binding recipe specialization in several tests (I thought I fixed this before, but it might have been undone by a merge...). I've fixed these to set the right property, and that also revealed a small bug in the interpreted version of the UnboxAddress binding, where we weren't properly reject heap segments. I've fixed that as well.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issues
 * [JDK-8301261](https://bugs.openjdk.org/browse/JDK-8301261): Add linker option for specifying uncaught exception handler
 * [JDK-8301249](https://bugs.openjdk.org/browse/JDK-8301249): Linker::upcalStub should accept linker options


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/777/head:pull/777` \
`$ git checkout pull/777`

Update a local copy of the PR: \
`$ git checkout pull/777` \
`$ git pull https://git.openjdk.org/panama-foreign pull/777/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 777`

View PR using the GUI difftool: \
`$ git pr show -t 777`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/777.diff">https://git.openjdk.org/panama-foreign/pull/777.diff</a>

</details>

JornVernee commented:
/solves https://bugs.openjdk.org/browse/JDK-8301249

mcimadamore commented on test/jdk/java/foreign/TestUpcallException.java:
```diff
@@ -37,22 +37,126 @@
+
+    public static class NonVoidUpcallRunner extends ExceptionRunnerBase {
+        public static void main(String[] args) throws Throwable {
+            MethodHandle handle = MethodHandles.identity(int.class);
```
I'm not 100% sure that building these complex method handle chains is better than just having two throwing methods in the code - one with zero arguments and one with one argument (int) ?

JornVernee replied:
Yeah, that seems simpler.

JornVernee commented on src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:
```diff
@@ -180,7 +181,10 @@ private static MethodHandle specializeUpcall(MethodHandle leafHandle, CallingSeq
             // where the caller is an upcall stub.
-            MethodHandles.Lookup defineClassLookup = MethodHandles.lookup().defineHiddenClass(bytes, true);
+            Thread.UncaughtExceptionHandler uncaughtExceptionHandler = callingSequence.uncaughtExceptionHandler();
+            MethodHandles.Lookup defineClassLookup = uncaughtExceptionHandler != null
```
The default handler would be the one installed on the thread when we encounter an exception. I don't think we can get that here, since someone might link the upcall stub, and then afterwards set the thread's default exception handler.

The handler specified as a linker option can be used to override that handler on a per upcall stub basis, if wanted.

mcimadamore replied:
I see - the "fallback handler" is a dynamic property (depends on the thread in which we're executing the catch block". Then I agree the code, as is now, makes sense.

JornVernee commented:
I've addressed review comments.

I wrote this code a while back, and at the time I had used `exactInvoker` in the tests to bypass the exception that you get when trying to link an upcall stub with a target method handle that provably throws an exception. I've now decided that a simpler way to deal with that issue is to just throw an unchecked exception. (So that's where the switch from `Throwable` -> `RuntimeException` comes from)

JornVernee commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -332,6 +333,19 @@ static Option isTrivial() {
+         * {@return a linker option that can be used to specify the uncaught exception handler that should be executed
+         *          if an exception is thrown, but not caught, during an upcall}
+         * <p>
+         * Note that using a custom exception handler will not prevent the VM from exiting in the case of an uncaught
```
Done

mcimadamore reviewed (approved):
I like the test changes - much simpler now!

JornVernee commented:
/integrate

dholmes-ora commented:
FTR I think this is not a good idea with regards to using the Thread's UEH. That is only supposed to be called as a thread terminates due to the uncaught exception.

JornVernee commented:
When we encounter an uncaught exception like this, the process will terminate, so using the thread's UEH seems appropriate?

dholmes-ora commented:
Not really no. Unless this only happens in thread's you create and control and know what the UEH does, then invoking an arbitrary thread's UEH is a bad idea. You have no idea what it may do.
